### PR TITLE
gh-96143: Improve perf profiler docs

### PR DESCRIPTION
**Pull request by erlend-aasland**
- Add What's New
- Rewrite NEWS entry to match What's New
- Document new sys APIs
- Fix formatting here and there
- Adjust wording a few places

* Issue: gh-96143

### CODE REVIEW

File: Doc/howto/perf_profiling.rst [hunk 1]
```diff
@@ -25,7 +25,7 @@ fly before the execution of every Python function and it will teach ``perf`` the
 relationship between this piece of code and the associated Python function using
 `perf map files`_.
 
-.. warning::
+.. note::
```
Comment (erlend-aasland) [inline Doc/howto/perf_profiling.rst:29]: IMO, warning is too much; a note is better suited.

File: Doc/library/sys.rst [hunk 2]
```diff
@@ -1524,6 +1524,35 @@ always available.
       This function has been added on a provisional basis (see :pep:`411`
       for details.)  Use it only for debugging purposes.
 
+.. function:: activate_stack_trampoline(backend, /)
+
+   Activate the stack profiler trampoline *backend*.
```
Comment (pablogsal) [inline Doc/library/sys.rst:1560]: Maybe link against the `perf` docs?

File: Doc/howto/perf_profiling.rst [hunk 3]
```diff
@@ -172,27 +180,22 @@ profiling mode by using the APIs in the :mod:`sys` module:
...
 For the best results, Python should be compiled with
 ``CFLAGS="-fno-omit-frame-pointer -mno-omit-leaf-frame-pointer"`` as this allows
 profilers to unwind using only the frame pointer and not on DWARF debug
-information. This is because as the code that is interposed to allow perf
+information. This is because as the code that is interposed to allow ``perf``
 support is dynamically generated it doesn't have any DWARF debugging information
 available.
 
-You can check if you system has been compiled with this flag by running:
+You can check if you system has been compiled with this flag by running::
```
Comment (erlend-aasland) [inline Doc/howto/perf_profiling.rst]:
```suggestion
You can check if your system has been compiled with this flag by running::
```

File: Doc/whatsnew/3.12.rst [hunk 4]
```diff
@@ -74,6 +74,14 @@ Important deprecations, removals or restrictions:
...
 ============
 
+* Add :ref:`perf_profiling` through the new command-line option
+  :option:`-X perf <-X>`, as well as the new
+  :func:`sys.activate_stack_trampoline`,
+  :func:`sys.deactivate_stack_trampoline`,
+  and :func:`sys.is_stack_trampoline_active` APIs.
+  (Contributed by Pablo Galindo and Christian Heimes
+  with contributions from Gregory P. Smith [Google] and Mark Shannon
+  in :gh:`96123`.)
```
Comment (pablogsal) [inline Doc/whatsnew/3.12.rst]:
```suggestion
  (Design by Pablo Galindo. Contributed by Pablo Galindo and Christian Heimes
  with contributions from Gregory P. Smith [Google] and Mark Shannon
  in :gh:`96123`.)
```

File: Doc/library/sys.rst (hunk 2 above)
Comment (erlend-aasland) [reply Doc/library/sys.rst:1560]: Yeah, more links :) What are the official perf docs? This, I presume: https://perf.wiki.kernel.org/index.php/Main_Page

File: Doc/howto/perf_profiling.rst [hunk 5]
```diff
@@ -147,19 +147,27 @@ Instead, if we run the same experiment with perf support activated we get:
...
 
-Enabling perf profiling mode
-----------------------------
+How to enable the ``perf`` profiling mode
+-----------------------------------------
 
-There are two main ways to activate the perf profiling mode. If you want it to be
-active since the start of the Python interpreter, you can use the `-Xperf` option:
+There are three ways to activate the ``perf`` profiling mode:
+using the :option:`-Xperf <-X>` command-line option,
```
Comment (pablogsal) [inline Doc/howto/perf_profiling.rst]: We could maybe make these bullet points?

File: Doc/library/sys.rst (hunk 2 above)
Comment (pablogsal) [reply Doc/library/sys.rst:1560]: Ah sorry, I meant to link against `Doc/howto/perf_profiling.rst`

Comment (erlend-aasland) [reply Doc/library/sys.rst:1560]: We should also link to those in the howto.

File: Doc/howto/perf_profiling.rst (hunk 5 above)
Comment (erlend-aasland) [reply Doc/howto/perf_profiling.rst]: Yes, that's nice. Added in 31c0eb819c2b951051e5cb12d497db01af28bcd4.

File: Doc/library/sys.rst (hunk 2 above)
Comment (erlend-aasland) [reply Doc/library/sys.rst:1560]: Added in 31c0eb819c2b951051e5cb12d497db01af28bcd4

Comment (erlend-aasland) [reply Doc/library/sys.rst:1560]: > Ah sorry, I meant to link against `Doc/howto/perf_profiling.rst`

Aha, that was already in place in the `seealso` just below ;) However, linking to the perf wiki is also nice.

File: Doc/howto/perf_profiling.rst [hunk 6]
```diff
@@ -51,11 +52,11 @@ For example, consider the following script:
     if __name__ == "__main__":
         baz(1000000)
 
-We can run perf to sample CPU stack traces at 9999 Hertz:
+We can run ``perf`` to sample CPU stack traces at 9999 Hertz::
```
Comment (vstinner) [inline Doc/howto/perf_profiling.rst:55]: For me, 1 kHz (1000 Hz) or 10 kHz (10 000 Hz) would be less surprising.

File: Doc/using/cmdline.rst [hunk 7]
```diff
@@ -535,12 +535,13 @@ Miscellaneous options
...
-   * ``-X perf`` to activate compatibility mode with the ``perf`` profiler.
+   * ``-X perf`` activates compatibility mode with the ``perf`` profiler.
      When this option is activated, the Linux ``perf`` profiler will be able to
      report Python calls. This option is only available on some platforms and
      will do nothing if is not supported on the current system. The default value
-     is "off". See also :envvar:`PYTHONPERFSUPPORT` and :ref:`perf_profiling`
-     for more information.
+     is "off". See :ref:`perf_profiling` for more details.
+     This option is equivalent to setting the environment variable
+     :envvar:`PYTHONPERFSUPPORT` to ``1``.
```
Comment (vstinner) [inline Doc/using/cmdline.rst]: It's not if -E or -I is used. I prefer to just add a reference to the env var.

File: Doc/using/cmdline.rst [hunk 8]
```diff
@@ -1036,8 +1037,13 @@ conflict.
 .. envvar:: PYTHONPERFSUPPORT
 
    If this variable is set to a nonzero value, it activates compatibility mode
-   with the ``perf`` profiler so Python calls can be detected by it. See the
-   :ref:`perf_profiling` section for more information.
+   with the ``perf`` profiler so Python calls can be detected by it.
+   This is exactly equivalent to setting :option:`-X perf <-X>`
```
Comment (vstinner) [inline Doc/using/cmdline.rst]: It's not equivalent if env vars are ignored. Why not just adding a link to the -X perf option without trying to say that it's *exactly* the same?

**Review by pablogsal — APPROVED** [review]
LGTM

Great work! 👍

**Review by vstinner — COMMENTED** [review]
I'm confused by "compatibility mode" wording whereas the env var is called "perf support". I would prefer to say that -X perf **enables support** for the Linux perf tool.

"Enable" support or "Add" support, not sure which one makes more send. Technically, the option *adds* something ;-)

The PyConfig member, the env var and the -X option have 3 different names which also confuse me. Would it make sense to rename the -X option to ``-X perfsupport`` to make it more consistent? If yes, I would suggest doing that in a separated PR.

The name "perf" is very generic and can mean many things. A newcomer might read it as: "use -X perf to get best performance" :-)

For me, "compatibility mode" sounds like "backward compatibility mode".

File: Doc/howto/perf_profiling.rst (hunk 6 above)
Comment (erlend-aasland) [reply Doc/howto/perf_profiling.rst:55]: I deliberately did not alter the example itself, but I agree that 9999 Hz is an unexpected number :)

File: Doc/using/cmdline.rst (hunk 8 above)
Comment (erlend-aasland) [reply Doc/using/cmdline.rst]: Note that several other environment variables have a similar description.

File: Doc/using/cmdline.rst (hunk 7 above)
Comment (erlend-aasland) [reply Doc/using/cmdline.rst]: I reverted this change.

**Comment by erlend-aasland** [comment]
> I'm confused by "compatibility mode" wording whereas the env var is called "perf support". I would prefer to say that -X perf enables support for the Linux perf tool.

+1

I've tried to address your review comments in 9c720233f0c61a748276efbac295702627edfdf3, @vstinner. Are you ok with those changes?

> The PyConfig member, the env var and the -X option have 3 different names which also confuse me. Would it make sense to rename the -X option to -X perfsupport to make it more consistent? If yes, I would suggest doing that in a separated PR.

+1

File: Doc/howto/perf_profiling.rst [hunk 9]
```diff
@@ -147,19 +148,29 @@ Instead, if we run the same experiment with perf support activated we get:
...
+----------------------------------------
 
-There are two main ways to activate the perf profiling mode. If you want it to be
-active since the start of the Python interpreter, you can use the `-Xperf` option:
+There are three ways to enable ``perf`` profiling support:
+
+* using the :option:`-Xperf <-X>` command-line option
+* using the :envvar:`PYTHONPERFSUPPORT` environment variable
+* using the :func:`sys.activate_stack_trampoline` and
+  :func:`sys.deactivate_stack_trampoline` APIs
```
Comment (ezio-melotti) [inline Doc/howto/perf_profiling.rst]:
```suggestion
* using the :func:`sys.activate_stack_trampoline` function
```
Technically `sys.deactivate_stack_trampoline` doesn't enable `perf` support, so maybe it shouldn't be mentioned here, especially since it's already mentioned below.

File: Doc/howto/perf_profiling.rst [hunk 10]
```diff
@@ -147,19 +148,29 @@ Instead, if we run the same experiment with perf support activated we get:
...
+
+If you want profiling to be active when you start the Python interpreter,
```
Comment (ezio-melotti) [inline Doc/howto/perf_profiling.rst]: This almost seems to imply that this is the only way to have it active from the start.  Wouldn't that be the case with the envvar too?

File: Doc/howto/perf_profiling.rst [hunk 11]
```diff
@@ -147,19 +148,29 @@ Instead, if we run the same experiment with perf support activated we get:
...
+use the :option:`-Xperf <-X>` option::
 
     $ python -Xperf my_script.py
 
-You can also set the :envvar:`PYTHONPERFSUPPORT` to a nonzero value to actiavate perf
-profiling mode globally.
+If you need to enable ``perf`` profiling support globally,
+set the environment variable :envvar:`PYTHONPERFSUPPORT`
+to a nonzero value.
```
Comment (ezio-melotti) [inline Doc/howto/perf_profiling.rst]: I find this sentence a bit weird.  If the envvar is set it's already implied that it will affect all the Python processes.  Also it might be better to suggest `1` instead of a "nonzero value" (maybe even show an example like `PYTHONPERFSUPPORT=1 python my_script.py`?).

File: Doc/howto/perf_profiling.rst [hunk 12]
```diff
@@ -147,19 +148,29 @@ Instead, if we run the same experiment with perf support activated we get:
...
 
-There is also support for dynamically activating and deactivating the perf
-profiling mode by using the APIs in the :mod:`sys` module:
+If you need to dynamically enable and disable ``perf`` profiling
+in response to a signal or other communication mechanisms with your process,
+use the :func:`sys.activate_stack_trampoline` and
+:func:`sys.deactivate_stack_trampoline` APIs:
```
Comment (ezio-melotti) [inline Doc/howto/perf_profiling.rst]: Why not just calling them functions?
```suggestion
:func:`sys.deactivate_stack_trampoline` functions:
```

Comment (ezio-melotti) [inline Doc/howto/perf_profiling.rst]: It seems to me that you briefly listed the 3 options, and then expanded them a bit afterwards.  The 3 paragraphs don't add much IMHO, so you could either integrate them in the list, or condense them in a single paragraph (something like "The :option:`-Xperf <-X>` option and the :envvar:`PYTHONPERFSUPPORT` allow you to enable ``perf`` profiling from the start, whereas the :func:`sys.activate_stack_trampoline` and :func:`sys.deactivate_stack_trampoline` functions allow you to enable and disable it dynamically.").

Comment (ezio-melotti) [inline Doc/howto/perf_profiling.rst]: Should we mention the priority?  I'm assuming it's function > -X > envvar.

File: Doc/whatsnew/3.12.rst [hunk 13]
```diff
@@ -74,6 +74,14 @@ Important deprecations, removals or restrictions:
 New Features
 ============
 
+* Add :ref:`perf_profiling` through the new command-line option
+  :option:`-X perf <-X>`, as well as the new
```
Comment (ezio-melotti) [inline Doc/whatsnew/3.12.rst]: What about the envvar?  Also in previous examples there was no space between `-X` and `perf` (not that it matters much, but we might as well be consistent).

File: Python/clinic/sysmodule.c.h [hunk 14]
```diff
@@ -1192,7 +1192,7 @@ PyDoc_STRVAR(sys_deactivate_stack_trampoline__doc__,
 "deactivate_stack_trampoline($module, /)\n"
 "--\n"
 "\n"
-"Dectivate the perf profiler trampoline.");
+"Dectivate stack profiler trampoline backend.");
```
Comment (ezio-melotti) [inline Python/clinic/sysmodule.c.h]: Does this deactivate the (only) active backend?  (and what happens if there are not active backends?)

There is some asymmetry in the docstring of the previous function, which activate the *backend* passed as an argument, and this function that deactivates *a* backend (the active one?).

File: Doc/howto/perf_profiling.rst (hunk 10 above)
Comment (gpshead) [reply Doc/howto/perf_profiling.rst]: -X and envvar are equivalent activators (same block of code). It's worth documenting which one overrides the other when both are specified though.  -Xperf with PYTHONPERFSUPPORT=0 in particular.

File: Doc/howto/perf_profiling.rst (hunk 12 above)
Comment (gpshead) [reply Doc/howto/perf_profiling.rst]: i'd simplify the wording even further to just say ", use :func:X and :func:Y." without a "the" or calling the APIs or functions.  People will infer that by name and click on the :func: generated links.

Comment (gpshead) [reply Doc/howto/perf_profiling.rst]: yes lets document that. per Python/initconfig.c `config_init_perf_profiling` the envvar is read first, followed by -Xperf being checked for.  So -Xperf will override PYTHONPERFSUPPORT=0.  _(realistically looking at the code, the environment variable is ignored when it is 0, the internal `config->perf_profiling= 1` only happens when the value is non-zero.  so =0 in the environment variable cannot be used for anything, It won't disable perf profiling within a PyConfig struct created by something embedding Python for example.  Is that a bug?  we'll figure that out later outside of this PR)_

File: Doc/whatsnew/3.12.rst (hunk 13 above)
Comment (gpshead) [reply Doc/whatsnew/3.12.rst]: yes, we should mention the environment variable here.

**Review by gpshead — CHANGES_REQUESTED** [review]
thanks for the edits!

File: Doc/howto/perf_profiling.rst (hunk 12 above)
Comment (ezio-melotti) [inline Doc/howto/perf_profiling.rst]:
```suggestion
use :func:`sys.activate_stack_trampoline` and
:func:`sys.deactivate_stack_trampoline`:
```

Comment (ezio-melotti) [reply Doc/howto/perf_profiling.rst]: I added a new suggestion that removes `the` and `APIs`.

File: Doc/howto/perf_profiling.rst [hunk 15]
```diff
@@ -147,19 +148,29 @@ Instead, if we run the same experiment with perf support activated we get:
...
-Enabling perf profiling mode
-----------------------------
+How to enable ``perf`` profiling support
+----------------------------------------
 
-There are two main ways to activate the perf profiling mode. If you want it to be
-active since the start of the Python interpreter, you can use the `-Xperf` option:
+There are three ways to enable ``perf`` profiling support:
+
+* using the :option:`-Xperf <-X>` command-line option
```
Comment (vstinner) [inline Doc/howto/perf_profiling.rst]: IMO ``-X perf`` is more readable than ``-Xperf``, I suggest replacing it in the doc like this:

```suggestion
* using the :option:`-X perf <-X>` command-line option
```

**Review by vstinner — COMMENTED** [review]
Ezio made many interesting comments. I will wait until they are addressed to review again the PR ;-)

File: Python/clinic/sysmodule.c.h (hunk 14 above)
Comment (pablogsal) [reply Python/clinic/sysmodule.c.h]: > Does this deactivate the (only) active backend?

It deactivates any backend. Right now is only 1

> (and what happens if there are not active backends?)

Then is a noop

> There is some asymmetry in the docstring of the previous function, which activate the backend passed as an argument, and this function that deactivates a backend (the active one?).

Hummm good point. Maybe we should clarify this a bit better on the docstring by mentioning "the active backend"

Comment (erlend-aasland) [reply Python/clinic/sysmodule.c.h]: > Maybe we should clarify this a bit better on the docstring by mentioning "the active backend"

+1

File: Doc/howto/perf_profiling.rst (hunk 12 above)
Comment (erlend-aasland) [reply Doc/howto/perf_profiling.rst]: I agree, that's better.

**Comment by erlend-aasland** [comment]
Sorry for the delay. I finally got around to try to address the review. Please take a look at the changes.

File: Doc/howto/perf_profiling.rst [hunk 16]
```diff
@@ -147,52 +148,61 @@ Instead, if we run the same experiment with perf support activated we get:
...
+----------------------------------------
 
-There are two main ways to activate the perf profiling mode. If you want it to be
-active since the start of the Python interpreter, you can use the `-Xperf` option:
+There are three ways to enable ``perf`` profiling support;
+the environment variable :envvar:`PYTHONPERFSUPPORT` and the
+:option:`-X perf <-X>` option allow you to enable perf profiling from the start,
+whereas the :func:`sys.activate_stack_trampoline` and
+:func:`sys.deactivate_stack_trampoline` functions allow you to enable and
+disable it dynamically.
```
Comment (ezio-melotti) [inline Doc/howto/perf_profiling.rst]: This sentence could be simplified.  In particular it mentions three ways of enabling `perf`, and while it technically lists three, it sounds like there are only two (envvar/-X or `sys` functions).

I would suggest rephrasing it as
```suggestion
``perf`` profiling support can either be enabled from the start using
the environment variable :envvar:`PYTHONPERFSUPPORT` or the
:option:`-X perf <-X>` option,
or dynamically using :func:`sys.activate_stack_trampoline` and
:func:`sys.deactivate_stack_trampoline`.
```
I don't like too much the two `or` though, so maybe the two sentences could be separated (or swapped, but that conflicts with the priority order of the next paragraph).
```suggestion
``perf`` profiling support can be enabled from the start using
the environment variable :envvar:`PYTHONPERFSUPPORT` or the
:option:`-X perf <-X>` option.
It is also possible to enable and disable it at runtime using
:func:`sys.activate_stack_trampoline` and :func:`sys.deactivate_stack_trampoline`.
```

File: Doc/howto/perf_profiling.rst [hunk 17]
```diff
@@ -147,52 +148,61 @@ Instead, if we run the same experiment with perf support activated we get:
...
-profiling mode globally.
+Example, using the environment variable::
 
-There is also support for dynamically activating and deactivating the perf
-profiling mode by using the APIs in the :mod:`sys` module:
+   $ PYTHONPERFSUPPORT=1
+   $ python script1.py
+   $ python script2.py
+   $ python script3.py
+   $ perf report -g -i perf.data
```
Comment (ezio-melotti) [inline Doc/howto/perf_profiling.rst]: This document doesn't seem to mention how `perf.data` is created and what it contains.  It would appear that is created automatically (with the `perf.data` name) whenever `perf` mode is enabled, however it is not clear if in this example the results of the three executions are combined, if they are separated within the file, or if every separate command will create a new `perf.data` that will override the previous one.

I'm also not sure if running ``perf`` on three different files like this is a realistic use case.  If not, maybe the following is better:
```suggestion
   $ PYTHONPERFSUPPORT=1 python script1.py
   $ perf report -g -i perf.data
```

File: Doc/library/sys.rst [hunk 18]
```diff
@@ -1555,6 +1555,38 @@ always available.
...
+   The only supported backend is `"perf"`.
```
Comment (ezio-melotti) [inline Doc/library/sys.rst]:
```suggestion
   The only supported backend is ``"perf"``.
```

File: Python/clinic/sysmodule.c.h [hunk 19]
```diff
@@ -1231,7 +1231,7 @@ PyDoc_STRVAR(sys_activate_stack_trampoline__doc__,
 "activate_stack_trampoline($module, backend, /)\n"
 "--\n"
 "\n"
-"Activate the perf profiler trampoline.");
+"Activate stack profiler trampoline *backend*.");
```
Comment (ezio-melotti) [inline Python/clinic/sysmodule.c.h:1234]: Should this mention that `"perf"` is the only option too?

Comment (erlend-aasland) [reply Python/clinic/sysmodule.c.h:1234]: I'd not bother with that in the docstring.

File: Doc/howto/perf_profiling.rst (hunk 17 above)
Comment (erlend-aasland) [reply Doc/howto/perf_profiling.rst]: I agree that this part can be improved. Since I do not know the answers to your questions from the top of my head, I need to actually test this on a Linux system (I've only got WSL Ubuntu instance at my hands right now, and I can't get perf up and running there).

**Comment by pablogsal** [comment]
@erlend-aasland can you fix the conflicts and I will land this? Thanks!

**Comment by erlend-aasland** [comment]
> @erlend-aasland can you fix the conflicts and I will land this? Thanks!

Will do! Thanks for the heads-up. I forgot about this PR :)

**Review by pablogsal — APPROVED** [review]
Thanks a lot for the fixes! ❤️ 

I plan to land this this week as all the feedback has been addressed but I will give some time in case someone wants to request some further changes.

**Review by gpshead — APPROVED** [review]
